backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/InventoryLineModularContractHandler.java

```diff
@@ -0,0 +1,191 @@
+import lombok.NonNull;
+import org.adempiere.exceptions.AdempiereException;
+import org.adempiere.util.lang.impl.TableRecordReference;
+import org.adempiere.warehouse.WarehouseId;
+import org.compiere.model.I_M_Inventory;
+import org.compiere.model.I_M_InventoryLine;
+import org.springframework.stereotype.Component;
+
+import java.util.Optional;
+import java.util.stream.Stream;
+
+@Component
+public class InventoryLineModularContractHandler implements IModularContractTypeHandler<I_M_InventoryLine>
+{
+	private static final AdMessageKey MSG_REACTIVATE_NOT_ALLOWED = AdMessageKey.of("de.metas.contracts.modular.impl.InventoryLineModularContractHandler.ReactivateNotAllowed");
+	private static final AdMessageKey MSG_DESCRIPTION = AdMessageKey.of("de.metas.contracts.modular.impl.InventoryLineModularContractHandler.Description");
+
+	private final IOrgDAO orgDAO = Services.get(IOrgDAO.class);
+	private final IMsgBL msgBL = Services.get(IMsgBL.class);
+	private final IFlatrateDAO flatrateDAO = Services.get(IFlatrateDAO.class);
+	private final IInventoryBL inventoryBL = Services.get(IInventoryBL.class);
+
+	private final ModularContractSettingsDAO modularContractSettingsDAO;
+
+	public InventoryLineModularContractHandler(@NonNull final ModularContractSettingsDAO modularContractSettingsDAO)
+	{
+		this.modularContractSettingsDAO = modularContractSettingsDAO;
+	}
+
+	@NonNull
+	@Override
+	public Class<I_M_InventoryLine> getType()
+	{
+		return I_M_InventoryLine.class;
+	}
+
+	@Override
+	public boolean applies(@NonNull final I_M_InventoryLine inventoryLine)
+	{
+		return FlatrateTermId.ofRepoIdOrNull(inventoryLine.getModular_Flatrate_Term_ID()) != null;
+	}
+
+	@Override
+	@NonNull
+	public Optional<LogEntryCreateRequest> createLogEntryCreateRequest(
+			@NonNull final I_M_InventoryLine inventoryLine,
+			@NonNull final FlatrateTermId modularContractId)
+	{
+		final I_M_Inventory inventory = inventoryBL.getById(InventoryId.ofRepoId(inventoryLine.getM_Inventory_ID()));
+
+		final InventoryId reversalId = InventoryId.ofRepoIdOrNull(inventory.getReversal_ID());
+		if (reversalId != null)
+		{
+			return Optional.empty();
+		}
+
+		final ModularContractSettings modularContractSettings = modularContractSettingsDAO.getByFlatrateTermIdOrNull(modularContractId);
+		if (modularContractSettings == null)
+		{
+			throw new AdempiereException("Invalid contract missing settings")
+					.appendParametersToMessage()
+					.setParameter("M_InventoryLine_ID", inventoryLine.getM_InventoryLine_ID())
+					.setParameter("ModularContractID", modularContractId);
+		}
+
+		final Optional<ModularContractTypeId> modularContractTypeId = modularContractSettings.getModuleConfigs()
+				.stream()
+				.filter(config -> config.isMatchingClassName(InventoryLineModularContractHandler.class.getName()))
+				.map(ModuleConfig::getModularContractType)
+				.map(ModularContractType::getId)
+				.findFirst();
+
+		final Quantity quantity = inventoryBL.getMovementQty(inventoryLine);
```

Comment:
**💡 can be done later**
if modularContractTypeId is empty, we can spare ourselves the additional work and return Optional.empty() right here.
This is also clearer IMO, and we can avoid the `modularContractTypeId.map` down the road, which is once again clearer

Reply:
done